### PR TITLE
Supporting memhavoc in Executable cex 

### PR DESCRIPTION
**Pull request by danblitzhou**
Recall definition of memhavoc: 
`void memhavoc(void *ptr, size_t sz)`
this tells seahorn to treat memory content starting at `ptr` with size in bytes of `sz` as non-deterministic value.

With change in https://github.com/seahorn/sea-dsa/pull/105, we can track shadow mem symbol of `ptr` and thus track SMT-solver value of non-deterministic memory content.

In CexHarness.cc:
- find and store shadow.mem Expr of `ptr` and value of `sz` for each memhavoc call
- `exprToMemSegment(ptr, sz, ctx, datalayout)`: use HexDump to convert `ptr` Expr to key value pairs of <address id, memory content block> where each block length is word size of current data layout; If Key value pairs only contains default value, put default content block in `n` size array where `n=ceil(size / block size)`, otherwise convert key value pairs into content block arrays (XXX: this is not working at the moment)
- store pointers to all generated memory content block array for memhavoc in a global variable, 
- synthesis memhavoc(ptr, sz):
  - get counter, increment counter like other nd functions
  - get pointer `src` to memory content block array from global array at counter index (`__seahorn_get_value_ptr`)
  - call `memcpy(ptr, src, sz)` to fill memory

### CODE REVIEW

File: lib/seahorn/Smt/HexDump.cc [hunk 1]
```diff
@@ -261,7 +263,7 @@ class Pairs {
     // value into the set ( with a blank index)
     if (m_set.size() == 0)
       m_set.emplace(mkTerm<std::string>("*", m_defaultValue->efac()),
-                    m_defaultValue);
+                    m_defaultValue, true);
```
Comment (danblitzhou) [inline lib/seahorn/Smt/HexDump.cc:266]: reverting since `isRepeated()` should only be relevant between "gaps"

File: lib/seahorn/CexHarness.cc [hunk 2]
```diff
@@ -199,6 +303,60 @@ std::unique_ptr<Module> createCexHarness(BmcTraceWrapper &trace,
...
+            if (auto *hPtrAllocInst = dyn_cast<Instruction>(hPtr)) {
+              hPtrE = trace.eval(loc, *hPtrAllocInst, true);
+            } else {
+              LOG("cex", errs() << "unhandled Value of memhavoc ptr: " << *hPtr
+                                << "\n");
+              continue;
+            }
+
+            FuncValueMap[CF].push_back(shadowMemV);
+            HavocPtrs.push_back(std::make_pair(hPtrE->arg(0), sizeE));
```
Comment (agurfinkel) [inline lib/seahorn/CexHarness.cc:356]: I think this silently assumes that memory is a struct. This is not always the case.
At any rate, comment what `->arg(0)` is for.

Comment (agurfinkel) [reply lib/seahorn/CexHarness.cc:340]: has this been addressed or is it obsolete?

**Comment by agurfinkel** [comment]
please clang-format so that formatting check stops failing

File: include/seahorn/Bmc.hh [hunk 3]
```diff
@@ -14,20 +21,66 @@
 
 namespace seahorn {
 using namespace expr;
+using ZModel_ref = std::shared_ptr<ZModel<EZ3>>;
```
Comment (agurfinkel) [inline include/seahorn/Bmc.hh:24]: I think our naming convention is not to mix camel and snake case. That is, either caps or underscore, but not both.

File: include/seahorn/Bmc.hh [hunk 4]
```diff
@@ -14,20 +21,66 @@
...
 
 namespace bmc_impl {
 /// true if I is a call to a void function
 bool isCallToVoidFn(const llvm::Instruction &I);
 /// computes an implicant of f (interpreted as a conjunction) that
 /// contains the given model
-void get_model_implicant(const ExprVector &f, ZModel<EZ3> &model,
-                         ExprVector &out, ExprMap &active_bool_map);
+template <typename model_ref>
```
Comment (agurfinkel) [inline include/seahorn/Bmc.hh:31]: minimize code that lives in the header file used by clients. 
`Bmc.hh` is included by clients of Bmc engines. Clients should not import implementation. 

Put implementation into another header file that is only included by `Bmc.cc`. Usual name is `BmcImpl.hh` or some similar convention

File: include/seahorn/Bmc.hh [hunk 5]
```diff
@@ -79,7 +132,8 @@ public:
   }
 
   /// Returns the BMC trace (if available)
-  virtual BmcTrace getTrace();
+  template <class Engine, class Model>
+  BmcTrace<Engine, Model> getTrace();
```
Comment (agurfinkel) [inline include/seahorn/Bmc.hh:136]: define local type for `BmcTrace`. For example
```
using BmcTraceTy = BmcTrace<Engine, Model>
```

File: include/seahorn/Bmc.hh [hunk 6]
```diff
@@ -117,10 +171,10 @@ public:
   std::vector<SymStore> &getStates() { return m_states; }
 };
 
+template <class Engine, class Model>
```
Comment (agurfinkel) [inline include/seahorn/Bmc.hh:174]: Move into `BmcImpl.hh` header. Eventually, we will make it a subclass of some general `BmcTrace` class. For now, moving into a header file will suffice.

File: include/seahorn/Expr/ExprMemMap.hh [hunk 7]
```diff
@@ -0,0 +1,174 @@
+/// Pointer Expr -> Symbolic Memory Map
```
Comment (agurfinkel) [inline include/seahorn/Expr/ExprMemMap.hh:1]: Is this new code or factored out from `HexDump`?

File: include/seahorn/MemSimulator.hh [hunk 8]
```diff
@@ -31,20 +31,20 @@ class MemSimulator {
   // -- start of internally allocated memory
   unsigned m_intMemStart;
```
Comment (agurfinkel) [inline include/seahorn/MemSimulator.hh:32]: `MemSimulator` is dead code. If it gets too hard to update it, we can remove it.

File: lib/seahorn/Bmc.cc [hunk 9]
```diff
@@ -130,205 +129,11 @@ raw_ostream &BmcEngine::toSmtLib(raw_ostream &out) {
   return m_smt_solver.toSmtLib(out);
 }
 
-BmcTrace BmcEngine::getTrace() {
+template <> BmcTrace<BmcEngine, ZModel_ref> BmcEngine::getTrace() {
```
Comment (agurfinkel) [inline lib/seahorn/Bmc.cc:132]: this template specialization is unexpected here. Probably issue somewhere else in the design. 
I expect not to have `template<>` in front of this method.

File: lib/seahorn/Bmc.cc [hunk 10]
```diff
@@ -130,205 +129,11 @@ raw_ostream &BmcEngine::toSmtLib(raw_ostream &out) {
...
-        out << " [" << (*dloc).getFilename() << ":" << dloc.getLine() << "]";
-      }
-      out << "\n";
-      out.resetColor();
-    }
-  }
-  out << "End trace\n";
-  return out;
+  return BmcTrace<BmcEngine, ZModel_ref>(
+      *this, std::make_shared<ZModel<EZ3>>(model));
```
Comment (agurfinkel) [inline lib/seahorn/Bmc.cc:136]: Worried about this line. What does it actually do? 
Create a model on the heap using a copy constructor and and making it shared?

This is quite different from the previous design that does not allocate on the heap. Why the change?

File: lib/seahorn/Bmc.cc [hunk 11]
```diff
@@ -439,6 +198,35 @@ void unsat_core(ZSolver<EZ3> &solver, const ExprVector &f, bool simplify,
     out.push_back(bind::fname(bind::fname(c))->arg(0));
 }
 
+template <>
```
Comment (agurfinkel) [inline lib/seahorn/Bmc.cc:201]: why `template<>`?

File: lib/seahorn/HornCex.cc [hunk 12]
```diff
@@ -95,7 +95,8 @@ using namespace llvm;
 namespace seahorn {
 
 template <typename O> class SvCompCex;
-static void dumpSvCompCex(BmcTrace &trace, std::string CexFile);
+static void dumpSvCompCex(BmcTrace<BmcEngine, ZModel_ref> &trace,
```
Comment (agurfinkel) [inline lib/seahorn/HornCex.cc:98]: Add a type for `BmcTrace<BmcEngine, ZModel_ref>` and use throughout. For example, call it `ZBmcTrace`

File: lib/seahorn/HornCex.cc [hunk 13]
```diff
@@ -302,7 +303,7 @@ bool HornCex::runOnFunction(Module &M, Function &F) {
   LOG("cex", errs() << "Validated CEX by BMC engine.\n";);
 
   // get bmc trace
-  BmcTrace trace(bmc.getTrace());
+  BmcTrace<BmcEngine, ZModel_ref> trace(bmc.getTrace<BmcEngine, ZModel_ref>());
```
Comment (agurfinkel) [inline lib/seahorn/HornCex.cc:306]: This looks wrong. `getTrace` should not need to have template parameters specified. It gets them from its class `bmc`

File: lib/seahorn/HornCex.cc [hunk 14]
```diff
@@ -483,7 +484,8 @@ static void debugLocToSvComp(const Instruction &inst, SvCompCex<O> &svcomp) {
   svcomp.edge(file, (int)dloc.getLine(), "");
 }
 
-static void dumpSvCompCex(BmcTrace &trace, std::string CexFile) {
+static void dumpSvCompCex(BmcTrace<BmcEngine, ZModel_ref> &trace,
```
Comment (agurfinkel) [inline lib/seahorn/HornCex.cc:487]: type for `BmcTrace`. see comment above.

File: lib/seahorn/SolverBmc.cc [hunk 15]
```diff
@@ -120,274 +118,11 @@ raw_ostream &SolverBmcEngine::toSmtLib(raw_ostream &out) {
   return out;
 }
 
-SolverBmcTrace SolverBmcEngine::getTrace() {
+template <>
```
Comment (agurfinkel) [inline lib/seahorn/SolverBmc.cc:121]: should not need `template<>`

**Review by agurfinkel — CHANGES_REQUESTED** [review]
Looking good. A few confusions on my part around templates.

let's fix it and merge as soon as possible

File: include/seahorn/Expr/ExprMemMap.hh (hunk 7 above)
Comment (danblitzhou) [reply include/seahorn/Expr/ExprMemMap.hh:1]: factored out from `HexDump`; cut short much of the layers of old design;
now its just:
`ExprMemMap` has `MemCellSet` and `MemExprVisitor`;
`MemCellSet` contains a set of `MemCell`s and functions for access

File: include/seahorn/MemSimulator.hh (hunk 8 above)
Comment (danblitzhou) [reply include/seahorn/MemSimulator.hh:32]: `HornCex.cc` is the only place that still uses it, what is our plan for that file?

File: lib/seahorn/Bmc.cc (hunk 10 above)
Comment (danblitzhou) [reply lib/seahorn/Bmc.cc]: I changed `m_model` from a reference to a shared pointer to match `Solver::model_ref`, which is a shared pointer of model implementation.

File: lib/seahorn/Bmc.cc (hunk 11 above)
Comment (danblitzhou) [reply lib/seahorn/Bmc.cc:200]: this is a specialization of template `Out` using llvm::raw_ostream

**Comment by danblitzhou** [comment]
Regarding `CexHarness`, right now I just copied some functions into a `impl` header file to make it work with template `BmcTrace` so that `horn-bmc-solver=smt-y2` produces cex harness. 
I have a more comprehensive refactor of cex harness generating system almost ready on top of changes in this PR, maybe it's better to put those changes in a separate PR.

**Comment by agurfinkel** [comment]
CI in azure times out after 1.5 hrs. Do the tests pass for you locally, or is that a genuine problem?

File: include/seahorn/MemSimulator.hh (hunk 8 above)
Comment (agurfinkel) [reply include/seahorn/MemSimulator.hh:32]: We can refactor anything we want. There are no other dependencies other than our code.

File: lib/seahorn/Bmc.cc [hunk 16]
```diff
@@ -130,205 +129,10 @@ raw_ostream &BmcEngine::toSmtLib(raw_ostream &out) {
...
-          out.changeColor(raw_ostream::CYAN);
-        out << " [" << (*dloc).getFilename() << ":" << dloc.getLine() << "]";
-      }
-      out << "\n";
-      out.resetColor();
-    }
-  }
-  out << "End trace\n";
-  return out;
+  return ZBmcTraceTy(*this, std::make_shared<ZModel<EZ3>>(model));
```
Comment (agurfinkel) [inline lib/seahorn/Bmc.cc:135]: This is potentially problematic. Double check ownership.

`model` from the `m_smt_solver` is already a smart pointer, wrapping it in another smart pointer is at the very least annoying. Plus the allocation is shifted to the heap...

File: lib/seahorn/Bmc.cc (hunk 11 above)
Comment (agurfinkel) [reply lib/seahorn/Bmc.cc:200]: add a comment that this is a specialization, just like the comment above

File: lib/seahorn/BmcPass.cc [hunk 17]
```diff
@@ -348,24 +348,24 @@ class BmcPass : public llvm::ModulePass {
...
-          errs() << "Analyzed Function:\n" << F << "\n";
-          errs() << "Trace \n";
-          BmcTrace trace(bmc.getTrace());
-          trace.print(errs());
-        });
-
     if (res) {
+      ZBmcTraceTy trace(bmc.getTrace());
+      LOG(
+          "cex", if (res) {
```
Comment (agurfinkel) [inline lib/seahorn/BmcPass.cc:354]: formatting. I usually do 
```
LOG("cex", { if... } );
```
Then clang-format works properly.

File: lib/seahorn/BmcPass.cc [hunk 18]
```diff
@@ -348,24 +348,24 @@ class BmcPass : public llvm::ModulePass {
...
       if (CexFileRef != "") {
         if (CexFileRef.endswith(".ll") || CexFileRef.endswith(".bc")) {
           auto &tli = getAnalysis<TargetLibraryInfoWrapperPass>();
           auto const &dl = F.getParent()->getDataLayout();
-          BmcTrace trace(bmc.getTrace());
-          BmcTraceWrapper trace_wrapper(trace);
-          dumpLLVMCex(trace_wrapper, CexFileRef, dl, tli.getTLI(F),
-                      F.getContext());
+          BmcTraceWrapper<ZBmcTraceTy> trace_wrapper(trace);
+          dumpLLVMCex<ZBmcTraceTy>(trace_wrapper, CexFileRef, dl, tli.getTLI(F),
```
Comment (agurfinkel) [inline lib/seahorn/BmcPass.cc:367]: `ZBmcTraceTy` should not be needed. If it is needed, then something is wrong with how it is declared.

File: lib/seahorn/BmcPass.cc [hunk 19]
```diff
@@ -397,27 +397,40 @@ class BmcPass : public llvm::ModulePass {
...
-    LOG(
-        "cex", if (res == solver::SolverResult::SAT) {
-          errs() << "Analyzed Function:\n" << F << "\n";
-          errs() << "Trace \n";
-          SolverBmcTrace trace(bmc.getTrace());
-          trace.print(errs());
-        });
-    // todo: support --cex mode (generate harness)
+    if (res == SolverResult::SAT) {
+      SolverBmcTraceTy trace(bmc.getTrace());
```
Comment (agurfinkel) [inline lib/seahorn/BmcPass.cc:414]: this whole `if` should be inside the `LOG` macro. If `LOG` is disabled, none of this code should be executed.

File: lib/seahorn/BmcPass.cc [hunk 20]
```diff
@@ -397,27 +397,40 @@ class BmcPass : public llvm::ModulePass {
...
+        errs() << "Trace \n";
+        trace.print(errs());
+      });
+      StringRef CexFileRef(HornCexFile);
+      if (CexFileRef != "") {
+        if (CexFileRef.endswith(".ll") || CexFileRef.endswith(".bc")) {
+          auto &tli = getAnalysis<TargetLibraryInfoWrapperPass>();
+          auto const &dl = F.getParent()->getDataLayout();
+          BmcTraceWrapper<SolverBmcTraceTy> trace_wrapper(trace);
+          dumpLLVMCex<SolverBmcTraceTy>(trace_wrapper, CexFileRef, dl,
```
Comment (agurfinkel) [inline lib/seahorn/BmcPass.cc:426]: `SolverBmcTraceTy` is not needed as specialization for `dumpLLVMVCex`

File: lib/seahorn/HornCex.cc [hunk 21]
```diff
@@ -327,11 +327,12 @@ bool HornCex::runOnFunction(Module &M, Function &F) {
...
   if (HornCexFileRef.endswith(".ll") || HornCexFileRef.endswith(".bc")) {
     const DataLayout &dl = M.getDataLayout();
-    std::unique_ptr<BmcTraceWrapper> traceW(new BmcTraceWrapper(trace));
+    std::unique_ptr<BmcTraceWrapper<ZBmcTraceTy>> traceW(
+        new BmcTraceWrapper<ZBmcTraceTy>(trace));
     if (memSim) {
       traceW.reset(new BmcTraceMemSim(*memSim));
     }
-    dumpLLVMCex(*traceW, HornCexFileRef, dl, tli, F.getContext());
+    dumpLLVMCex<ZBmcTraceTy>(*traceW, HornCexFileRef, dl, tli, F.getContext());
```
Comment (agurfinkel) [inline lib/seahorn/HornCex.cc:335]: no need for `ZBmcTraceTy`

**Review by agurfinkel — CHANGES_REQUESTED** [review]

File: lib/seahorn/Bmc.cc (hunk 16 above)
Comment (danblitzhou) [reply lib/seahorn/Bmc.cc:135]: `m_smt_solver.getModel()` returns a `ZModel<EZ3>` object/reference, which is why I wanted to wrap it around `ZModelRef` in the first place; perhaps I should initialize `ZBmcTraceTy::m_model` with the pointer of `m_smt_solver.getModel()` directly?

Comment (agurfinkel) [reply lib/seahorn/Bmc.cc:135]: ZModel is a smart pointer. You can copy it by value.

**Review by agurfinkel — APPROVED** [review]
